File: python/day7.py
from utilities import read_lines
from collections import defaultdict, deque
from typing import List, Dict


class RegulationGraph:
    def __init__(self, regulations: List[str]):
        self.parent_bags = defaultdict(list)
        self.inner_bags = defaultdict(list)
        for regulation in regulations:
            self.read_regulation(regulation)
        self.mem = {}

    def read_regulation(self, regulation: str):
        # Ignore the `.` at the end of line
        regulation = regulation[:-1]

        # Split into the outer bags and inner bags
        parts = regulation.split(' contain ')

        # Ignore the word ` bags`
        outer_bag = parts[0][:-5]

        inner_bags = parts[1].split(', ')
        for bag in inner_bags:
            if 'no other' in bag:
                continue
            color_start = bag.find(' ') + 1
            color_end = bag.rfind(' ')
            color = bag[color_start:color_end]

            self.parent_bags[color].append(outer_bag)

            number_of_inner_bag = int(bag[:color_start])
            self.inner_bags[outer_bag].append((number_of_inner_bag, color))

    def count_outermost_bag(self, color):
        s = set()
        q = deque()
        q.append(color)
        while len(q) != 0:
            color = q.popleft()
            for parent in self.parent_bags[color]:
                if parent not in s:
                    s.add(parent)
                    q.append(parent)
        return len(s)

    def count_bag_inside(self, color):
        mem = self.mem

        def helper(color):
            if mem.get(color) is None:
                if len(self.inner_bags[color]) == 0:
                    mem[color] = 0
                else:
                    ans = 0
                    for inner_bag in self.inner_bags[color]:
                        ans += inner_bag[0]
                        ans += inner_bag[0] * helper(inner_bag[1])
                    mem[color] = ans
            return mem.get(color)
        return helper(color)


def part_1(graph: RegulationGraph):
    return graph.count_outermost_bag('shiny gold')


def part_2(graph: RegulationGraph):
    return graph.count_bag_inside('shiny gold')


if __name__ == '__main__':
    print('Day 7')
    data = read_lines('../inputs/day7.txt')
    graph = RegulationGraph(data)
    print('\tPart 1: {}'.format(part_1(graph)))
    print('\tPart 2: {}'.format(part_2(graph)))

File: rust/src/day7.rs
use std::collections::{HashMap, HashSet, VecDeque};

use crate::utilities::read_lines;

struct Graph {
    parent_bags: HashMap<String, Vec<String>>,
    children_bags: HashMap<String, Vec<(usize, String)>>,
}

impl Graph {
    fn new(regulations: Vec<String>) -> Graph {
        let mut parent_bags: HashMap<String, Vec<String>> = HashMap::new();
        let mut children_bags: HashMap<String, Vec<(usize, String)>> = HashMap::new();

        let mut read_regulation = |regulation: String| {
            let mut parts = regulation.split(" contain ");
            let outer_bag_ = parts.next().unwrap();
            let outer_color_end = outer_bag_.rfind(' ').unwrap();
            let outer_bag = &outer_bag_[..outer_color_end];
            let inner_bags = parts.next().unwrap().split(", ");
            for bag in inner_bags {
                if bag.contains("no other") {
                    continue;
                }
                let color_start = bag.find(' ').unwrap() + 1;
                let color_end = bag.rfind(' ').unwrap();
                let color = &bag[color_start..color_end];

                let parent = parent_bags.entry(color.to_string()).or_default();
                parent.push(outer_bag.to_string());

                let num_of_bags: usize = bag[..color_start - 1].parse().unwrap();
                let children = children_bags.entry(outer_bag.to_string()).or_default();

                children.push((num_of_bags, color.to_string()));
            }
        };
        for regulation in regulations {
            read_regulation(regulation);
        }
        Graph {
            parent_bags,
            children_bags,
        }
    }

    fn count_outermost_bag(&self, color: &str) -> usize {
        let mut set = HashSet::new();
        let mut queue = VecDeque::new();
        queue.push_back(color);
        while let Some(color) = queue.pop_front() {
            if let Some(parents) = self.parent_bags.get(color) {
                for parent in parents {
                    if set.insert(parent) {
                        queue.push_back(parent);
                    }
                }
            }
        }
        set.len()
    }

    fn count_bag_inside(&self, color: &str) -> usize {
        let mut mem = HashMap::new();
        self.count_bag_helper(color, &mut mem)
    }

    fn count_bag_helper(&self, color: &str, mem: &mut HashMap<String, usize>) -> usize {
        if let Some(value) = mem.get(color) {
            return *value;
        }
        match self.children_bags.get(color) {
            None => 0,
            Some(children) => {
                let mut ans = 0;
                for child in children {
                    ans += child.0;
                    ans += child.0 * self.count_bag_helper(child.1.as_str(), mem);
                }
                mem.insert(color.to_string(), ans);
                ans
            }
        }
    }
}

fn part_1(graph: &Graph) {
    println!("\tPart 1: {}", graph.count_outermost_bag("shiny gold"));
}

fn part_2(graph: &Graph) {
    println!("\tPart 2: {}", graph.count_bag_inside("shiny gold"));
}

pub fn main() {
    println!("Day 7");
    let graph = Graph::new(read_lines("../inputs/day7.txt"));
    part_1(&graph);
    part_2(&graph);
}
